### ISSUE TITLE
Future Task Rejections after updating from gfc-kinesis

### ISSUE THREAD

**Issue by gmalouf** [gilt/gfc-aws-kinesis#24]
I noticed after upgrading from the original gfc-kinesis client that the new one creates it's own executor and in-general seems to behave a lot differently.  Our web applications do a decent number of transactions/second; for each one we are adding a single message to our Kinesis stream.  What we are seeing frequently are 'TaskRejectionExceptions' from the Future creation in the `publishBatch` function.  

While I can (and am) increasing the thread pool size up from the default of 8, I'm wondering if there is a better way I should be using the library than via the `publishBatch` function for my use-case.  Is there a different hook-method I should be using to quickly submit my single messages to Kinesis?

**Comment by gmalouf** [comment]
After digging in more, I think the use of a SynchronousQueue for the ThreadPoolExecutor is the issue.  My traffic is bursty (particularly at initialization), so always requiring the request to send immediately would require configuring a much larger thread pool than actually necessary...

**Comment by andreyk0** [comment]
Hi!
IIRC ThreadPoolExecutor creation was a cut'n'paste from `ExecutorService.newCachedThreadPool()`. I think the point was to give it a non-default thread builder. That seems to be part of the API now on java8, maybe it weren't or maybe I'd missed it but that's where that came from.

AmazonKinesisAsyncClient is actually a synchronous implementation if you peek inside. At least the version we have in the project. The main diff there seems to be that they use a thread pool size with 50 threads by default.

IDK how much is "decent" :) We don't have high transactional load. Recent history seems to be at around 70 R/s per JVM. Defaults work for that.
That said we don't publish a batch per incoming message. We aggregate a few transactions in memory before publishing a larger kinesis batch, mainly as a cost saving measure.

Given that underlying implementation is fundamentally synchronous, I imagine your target number of threads will, too, depend on the kinesis RPC call latency and your target throughput, so, tweaking thread pool is not 'wrong' and, obviously, if you can do micro-aggregation in memory you can reduce RPC overhead and costs.
Again, in this particular example service we aggregate for up to 1sec or up to 100 records, whichever comes first.

Here's our batcher implementation from the project, can move into the public lib if that's useful.


```scala
package util

import java.util.concurrent.atomic.AtomicReference
import java.util.{Timer, TimerTask}

import com.gilt.gfc.logging.Loggable

import scala.annotation.tailrec
import scala.concurrent.ExecutionContext
import scala.concurrent.duration.FiniteDuration
import scala.util.control.NonFatal


/** Batches multiple 'one-at-a-time' calls into a single batch. */
trait Batcher[R] {

  /** Add a single record to batch. */
  def add(r: R): Unit

  /** Flush all outstanding records. */
  def flush(): Unit

  /** Flush outstanding records and shutdown background tasks.
    * Batcher instance can not be used after this call.
    */
  def shutdown(): Unit
}


object Batcher {

  /** Creates a batcher.
    * This should be thread-safe.
    *
    * @param name                    for logging
    * @param maxOutstandingCount     a batch will be submitted when record count reaches this number, it is assumed to be relatively small, see implementation of flush() for details
    * @param maxOutstandingDuration  we'll submit a batch with this periodicity regardless of the outstanding record count
    * @param submitBatch             what to do with a batch of records, hopefully an asynchronous operation (fast call), must be thread-safe
    * @param executor                where to run background task, safe to use global if submitBatch is non-blocking
    * @tparam R                      record type parameter
    * @return                        constructed batcher
    */
  def apply[R]( name: String
              , maxOutstandingCount: Int
              , maxOutstandingDuration: FiniteDuration
             )( submitBatch: (Iterable[R]) => Unit
             )( implicit executor: ExecutionContext
             ): Batcher[R] = {

    new BatcherImpl[R](
      name,
      maxOutstandingCount,
      maxOutstandingDuration,
      submitBatch,
      executor
    )
  }
}


private[util] final
class BatcherImpl[R] (
  name: String,
  maxOutstandingCount: Int,
  maxOutstandingDuration: FiniteDuration,
  submitBatch: (Iterable[R]) => Unit,
  executor: ExecutionContext
) extends Batcher[R]
     with Loggable {

  private[this]
  val emptyBatch = (0, List.empty[R])

  private[this]
  val currentBatch = new AtomicReference(emptyBatch)

  @volatile
  private[this]
  var isRunning = true

  // Flush buffer periodically
  val flushTask =  new TimerTask {
    override def run(): Unit = {
      if (isRunning) {
        flush()
      }
    }
  }

  val timer = new Timer
  timer.scheduleAtFixedRate(flushTask, maxOutstandingDuration.toMillis, maxOutstandingDuration.toMillis)


  /** Add a single record to batch. */
  @tailrec
  override
  def add(r: R): Unit = {
    require(isRunning, s"${name} batcher is shutting down, can not add any more records.")

    val b@(batchSize, records) = currentBatch.get()
    val b1@(newBatchSize, newRecords) = (batchSize+1, r :: records)

    if (newBatchSize >= maxOutstandingCount) {
      if (currentBatch.compareAndSet(b, emptyBatch)) {
        safeSubmitBatch(newRecords)
      } else {
        add(r) // retry
      }
    } else {
      if (!currentBatch.compareAndSet(b, b1)) {
        add(r) // retry
      }
    }
  }

  /** Flush all outstanding records. */
  @tailrec
  override
  def flush(): Unit = {
    val b@(_, records) = currentBatch.get()
    if (currentBatch.compareAndSet(b, emptyBatch)) {
      safeSubmitBatch(records)
    } else {
      flush() // retry
    }
  }

  override
  def shutdown(): Unit = {
    isRunning = false
    timer.cancel()
    flush()
  }

  private[this]
  def safeSubmitBatch(records: List[R]): Unit = {
    if (!records.isEmpty) {
      try {
        // Reverse a list to preserve original order of elements.
        // This may not be strictly required but just in case.
        // Obviously, batch size is assumed to be relatively small so that this doesn't matter.
        // For large collections Vector.:+ would be better but for small batches linked list
        // should be fine and might even lead to less contention in add() method (prepend is very quick).
        submitBatch(records.reverse)
      } catch {
        case NonFatal(e) =>
          error(s"Failed to flush ${name} batch: ${e.getMessage}", e)
      }
    }
  }
}
```

**Comment by gmalouf** [comment]
Thank you for the detailed response Andrey!  I actually think your batch
client implementation would be quite useful for others, as recording web
transaction data that comes in piece-by-piece is a pretty common use case.

Having said that, my current work-around is either to configure a much
larger thread pool than needed under normal load or to catch those
exceptions and retry up to n times.  If the thread pool itself could be
overwritten, I would use a bounded queue to allow for some back pressure
during bursty periods (makes me interested to see how your Akka streams
impl works)...I'll consider forking to get this behavior as it appears to
be the easiest change for us to get this done.

**Comment by gheine** [comment]
@andreyk0 would be great if you could add this to gfc-util (or maybe gfc-concurrent?) as it's generic and not dependent on anything AWS specific.

**Comment by andreyk0** [comment]
Would this help tuning thread pool? https://github.com/gilt/gfc-aws-kinesis/pull/26

**Comment by gmalouf** [comment]
Yes, it definitely would.  The synchronous queue doesn't allow for handling
'bursty' traffic well without re-implementing with the batch style that you
described to me earlier @andreyk0.

**Comment by gmalouf** [comment]
Thank you for looking at this!

**Comment by andreyk0** [comment]
NP. Batcher PR: https://github.com/gilt/gfc-concurrent/pull/19

**Comment by gheine** [comment]
^^ released in gfc-concurrent  v0.3.6
